static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,152 @@
+
+### Logical Components
+
+There are five main components to Vgaterm:
+
+* VGA Timing
+* CPU Pixel Output
+* Displaying a Pixel Byte
+* Digital to Analog Converters (DAC)
+* IO control and video buffer updates
+
+{% include figure.html image="/assets/images/vgaterm-logical-diagram.jpg" caption="Vgaterm logical diagram. Details were omitted to show the high level units" width=800 %}
+
+Repositories containing all our code and schematics:
+* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
+* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
+* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)
+
+### VGA Timing
+
+VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the analog color signals to actually dispaly a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).
+
+:----:|:-----:
+![](/assets/images/vgaterm-timing-front.jpg) | ![](/assets/images/vgaterm-timing-back.jpg)
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. This is the most important software role as any jitter in timing can cause display artifacts.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+This was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
```

Comment:
```suggestion
Our "chasing-the-beam" FIFO approach was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
```

(and then also make sure to describe the "chasing the beam" bit here or earlier)

Reply:
ha is this just so I say chasing the beam?

Reply:
In that RTL class I took, I learned (via, like, offhand kind of references) that putting FIFOs between clock domains is the standard approach. So "to cross the clock domains, we cleverly used a FIFO" reads a little like "to determine the area of the rectangle, we cleverly multiplied the length times the width": it's extremely clever, which is why 1) we stumbled into it from first principles, and 2) everyone uses it so much that it's kind of banal. 

I think there ought to be just a little more detail as to what was clever/constraining about our approach: what we were reacting to (the price of memory & difficulty in constructing a full off-chip VRAM bank, especially on breadboards?), and/or a little more flavor as to how difficult the challenge of precise hardware/software synchronization first appeared (can we even make the software be cycle-precise? wait, are the clock cycles even precisely the same between chips? etc...) and thus how clever we felt when we said "wait, let's buffer it!".

Both is good too—but you gotta save something for the future posts, yeah?